chrome/content/zotero/xpcom/utilities.js

```diff
@@ -144,6 +144,238 @@ var CSL_TYPE_MAPPINGS = {
+			uri,
+			{
+				successCodes: [200],
+				headers: {
+					'Content-Type': 'application/json'
+				},
+				body: JSON.stringify(item.toJSON())
+			}
+		);
+		
+		return JSON.parse(res.response);
+	},
+	
+	/**
+	 * Translate an item from a given URL
+	 *
+	 * @param {String} url
+	 * @return {Promise<Object|null>}
+	 */
+	"getItemFromURL": async function (url) {
+		let doc = (await Zotero.HTTP.processDocuments(url, doc => doc))[0];
+		
+		let translate = new Zotero.Translate.Web();
+		translate.setDocument(doc);
+		
+		let translators = await translate.getTranslators();
+		if (!translators.length) {
+			return null;
+		}
+		
+		translate.setTranslator(translators[0]);
+		let options = {
+			libraryID: false
+		};
+		
+		let newItems;
+		try {
+			newItems = await translate.translate(options);
+			if (!newItems.length) {
+				return null;
+			}
+		}
+		catch (e) {
+			Zotero.logError(e);
+			return null;
+		}
+		
+		return newItems[0];
+	},
+	
+	/**
+	 * Get an item by identifier
+	 *
+	 * @param {Object} identifier
+	 * @return {Promise<Object|null>}
+	 */
+	"getItemFromAPI": async function (identifier) {
+		let translate = new Zotero.Translate.Search();
+		translate.setIdentifier(identifier);
+		
+		let translators = await translate.getTranslators();
+		translate.setTranslator(translators);
+		let newItems;
+		try {
+			newItems = await translate.translate({
+				libraryID: false
+			});
+		}
+		catch (e) {
+			Zotero.logError(e);
+			return null;
+		}
+		return newItems[0];
+	},
+	
+	/**
+	 * Update item metadata
+	 *
+	 * @param {Zotero.Item} item
+	 * @return {Promise<bool>}
+	 */
+	"updateMetadata": async function (item) {
+		// Get an identifier from the item
+		let identifier = Zotero.Utilities.getItemIdentifier(item);
+		if (!identifier) {
+			// If no identifier is found, use other item fields to resolve it
+			let resolvedIdentifier = await Zotero.Utilities.resolveIdentifier(item);
+			// For now, resolve only DOI identifiers
+			if (resolvedIdentifier.DOI) {
+				identifier = resolvedIdentifier;
+			}
+		}
+		
+		let newItem = null;
+		
+		if (identifier) {
+			let item1 = await Zotero.Utilities.getItemFromAPI(identifier);
+			let item2 = null;
+			
+			if (identifier.DOI) {
+				item2 = await Zotero.Utilities.getItemFromURL('https://dx.doi.org/' + identifier.DOI);
+			}
+			
+			// If both items are translated
+			if (item1 && item2) {
+				// Combine item1 and item2
+				for (let key in item2) {
+					if (Zotero.Utilities.fieldIsValidForType(key, item1.itemType)) {
+						// Add all item2 fields that doesn't exist in item1
+						if (!item1[key] && item2[key]) {
+							item1[key] = item2[key];
+						}
+					}
+				}
+				newItem = item1;
+			}
+			// If only item1 is translated
+			else if (item1) {
+				newItem = item1;
+			}
+			// If only item2 is translated
+			else if (item2) {
+				newItem = item2;
+			}
+		}
+		
+		// Translate by URL if translation by identifier failed
+		if (!newItem && item.getField('url')) {
```

Comment:
It's not strictly necessary, and would be a little annoying to do, but technically we should avoid trying the URL if we already tried the DOI above. We don't have to do that in the initial implementation, but should at least add a TODO to skip URLs we tried from resolved DOIs.

Reply:
Well, this is complicated because the resolved URL only exists in `Zotero.HTTP.request`. Unless the item was successfully translated and the translator set a URL field, but in this case we already have an item.

Reply:
The URL can exist outside of `Zotero.HTTP.request` if calling code handles redirects manually, like we're doing for PDF retrieval. See https://github.com/zotero/zotero/blob/30291328ed0469a28118e7e822c3a1d3e60e8c26/chrome/content/zotero/xpcom/attachments.js#L1691-L1729 for an example (ignoring some of the extra stuff in there).

We would still need to get the URL out of `getItemFromURL()` — e.g., with an optional `onRedirectURL(url)` callback or something like that.

Reply:
So you are saying that we should add a `beforeRequest` callback option to `Zotero.HTTP.request`? That wouldn't be hard, but we are calling it through `Zotero.HTTP.processDocuments` therefore we have to add the same option parameter to this function. And also we have to workaround this: https://github.com/mrtcode/zotero/blob/ec5424d5ae5e4697ed5ecafedb04f4a809314a26/chrome/content/zotero/xpcom/http.js#L890-L896

Reply:
No, we wouldn't need to modify `request()` or `processDocuments()`. We could just do a `request()` with `followRedirects: false` and `responseType: 'document'`, call the callback on each URL, and use `wrapDocument()` like `processDocuments()` does on the final response.

But it's actually even simpler than that, I think. We really just care about the final resolved URL, which we can get from the document returned from `processDocuments()`. So `translateURL()` could just take an `onResponseURL()` callback that was passed the final URL from that document, whether or not it was translatable.

(The alternative would be to not worry about keeping the same response signature the same and return an object with a 'url' property and a 'data' property (which would sometimes be null), but you could imagine various functions that perform HTTP requests taking notification callbacks to monitor/abort/etc. the underlying HTTP requests, so I'm OK with the callback approach.)

Reply:
I agree that the final URL is enough to have, but the problem is that if HTTP request fails and an exception is thrown we don't know what was the final URL which failed. Therefore if the original URL was a doi.org URL, we will likely query it again and that fail again (because item can have the same doi.org URL in URL field and also we are making the same URL by adding DOI to doi.org URL).

Reply:
I started to think that maybe retrying a URL that was already requested through resolved DOI isn't as big problem as we're thinking. On the one hand it would be good to avoid querying the same URL two times, but on the other hand, it depends on the reason why it failed. For example if `Zotero.HTTP.request` failed with `404` error then yes, no need to query again, but if it failed because of [browser offline](https://github.com/mrtcode/zotero/blob/ec5424d5ae5e4697ed5ecafedb04f4a809314a26/chrome/content/zotero/xpcom/http.js#L172) error or something else that isn't related with the specific publisher web page problem, then retrying isn't that bad. So for now I'm just leaving this with "TODO".